client/app/lib/redux/modules/sidebar/stacks.coffee

```diff
@@ -0,0 +1,306 @@
+            handleRoute '/IDE'
+            return resolve result
+
+          , followEvents = no
+  }
+
+
+reloadIDE = (machineSlug) ->
+  console.log 'machine Slug ', machineSlug
+  { computeController } = kd.singletons
+  computeController.reloadIDE machineSlug
+
+
+reinitStack = (stack, machines, template) -> (dispatch) ->
+
+  return  unless template or not stack
+
+  { computeController, appManager } = kd.singletons
+
+  series([
+    (next) -> dispatch(destroyStack stack, machines, 'reinitStack').then ->
+      next()
+    (next) ->
+      initializeStack(template)(dispatch)
+  ])
+
+
+
+
+# selectors
+
+myStackTemplates = createSelector(
+  bongo.all 'JStackTemplate'
+  (templates) -> _.pickBy templates, (template) -> template.originId is whoami()._id
+)
+
+myStacks = createSelector(
+  bongo.all 'JComputeStack'
+  (stacks) -> _.pickBy stacks, (stack) -> stack.originId is whoami()._id
+)
+
+privateStackTemplates = createSelector(
+  myStackTemplates
+  (templates) -> templates and _.values(templates).filter (template) -> template.accessLevel is 'private'
+)
+
+teamStackTemplates = createSelector(
+  myStackTemplates
+  (templates) -> templates and _.values(templates).filter (template) -> template.accessLevel is 'group'
+)
+
+draftStackTemplates = createSelector(
+  myStacks
+  myStackTemplates
+  (stacks, templates) ->
+    return null  unless stacks and templates
+
+    baseStackIds = _.values(stacks).map (s) -> s.baseStackId
+    _.pickBy templates, (template) -> not (template._id in baseStackIds)
+)
+
+
+sidebarStacks = createSelector(
+  myStacks
+  draftStackTemplates
+  (stacks, templates) ->
+    return null  unless stacks and templates
+
+    _.merge stacks, templates
+)
+
+stacksAndMachines = createSelector(
+  sidebarStacks
+  bongo.all 'JMachine'
+  (stacks, machines) ->
+    return null  unless stacks and machines
+
+    return _.mapValues stacks, (stack, key) ->
+      return stack.machines
+        .map (machineId) -> machines[machineId]
+        .filter Boolean
+)
+
+stacksAndTemplates = createSelector(
+  sidebarStacks
+  myStackTemplates
+  (stacks, templates) ->
+    return null unless stacks and templates
+
+    return _.mapValues stacks, (stack, key) ->
+      if stack.baseStackId
+        templates[stack.baseStackId]
+      else stack
+)
+
+stacksAndCredential = createSelector(
+  sidebarStacks
+  bongo.all 'JCredential'
+  (stacks, credentials) ->
+    return _.mapValues stacks, (stack, key) ->
+      provider = stackProvider stack
+      return _.values(credentials)
+        .filter (credential) ->
+          stack.credentials?["#{provider}"]?.first is credential.identifier
+)
+
+stackProvider = (stack) ->
+
+  { config, credentials } = stack
+  return 'managedVM'  unless config
+  { requiredProviders } = config
+  for selectedProvider in requiredProviders
+    break  if selectedProvider in ['aws', 'vagrant']
+
+  selectedProvider ?= (Object.keys credentials ? { aws: yes }).first
```

Comment:
`? {aws.yes}` part will never work since `Object.keys` will return an array no matter what. So checking the first element would be a better approach here:

```coffee
[key] = Object.keys credentials
selectedProvider ?= key ? 'aws'
```